### ISSUE TITLE
[API Proposal]: Allow `Control.Invoke` on `Binding`

### ISSUE THREAD

**Issue by zeh-almeida** [dotnet/winforms#8582]

**Comment by KlausLoeffelmann** [comment]
One concern that I have here is that 98% of the binding scenarios in WinForms are not done via code but via the WinForms Designer UI. We would need to figure out a way in the Designer to allow the user to use the new feature interactively. That would also mean changes to the CodeDOM serialization.

One alternative way to address this could be an option in the Designer for a Form/UserControl to use Invoke when necessary. We could implement that on the Designer-side as a Design-Time property for the Form or UserControl. (`UseAutoBindingInvoke`).

That said, _if_ we would need to change the signature of `Binding` and in the view of the new Command Binding feature in WinForms, we should also take feedback into consideration, where people want a simplified way to handle Converters, preferably based on an Interface we'd need to add. Converters in WinForms binding are possible right now alright, but very awkwardly to hook up (namely via events for each binding item). I could see something like `IBindingConverter` similar to `IValueConverter` in WPF but with its naming adapted to WinForms to avoid confusion.

To simplify this, we could pass a Converter to the Binding class in addition, which would lead to internally wiring up the required events, running the passed converters when needed. 

_If_ we need to change `Binding` in more regards, we should do it in one go.

@merriemcgaw FYI.

**Comment by KlausLoeffelmann** [comment]
And of course, FYI @russkie, if time allows! 😸

**Comment by RussKie** [comment]
Thank you for tagging me.

My first reaction was that you'd likely want pretty much all bindings to go via the UI thread. With that in mind, the API signature change isn't required - you only would need to update the implementation (breaking docs notifications and all that entails).
May be provide an opt-out mechanism, _if_ necessary - through a feature switch or an API...

**Comment by zeh-almeida** [comment]
@RussKie so instead of having a new constructor like the one I made, are you suggesting I add a property with public get and set?

If that is correct, the behavior could be controlled at any time after instantiation, which sounds easier to manage on direct code but I don't know if that translates well to generator use.

Of course, we could have both options as well, I just want to make it clear because I can adjust my PR

**Comment by kirsan31** [comment]
Guys, may be I missed something, but I can't get core point here :( The code from 1 post is wrong even **without binding**:

```cs
var thread = new Thread(() =>
{
    textBox.Text = "Updated test text"; // System.InvalidOperationException
});

thread.Start();
```

🤷‍♂️

**Comment by zeh-almeida** [comment]
@kirsan31 the binding goes both ways, as you can see by the `Assert` calls made.
This code actually came from the PR I opened, so I am sure it works. Of course, I can always adjust it, no problem.

the `System.InvalidOperationException` happens on the binding because it does not use `Control.Invoke` and with the adjustments I propose, you may be able to avoid it entirely.

I hope I answered your questions, Please let me know if you have more

**Comment by kirsan31** [comment]
> the binding goes both ways

Yea and from 1 post it's unclear about which way we are talking here. From the text I made assumption that from _object to control_. And it's ok in principle, but in common case you can change your bind object as your want in any thread and will see no exceptions and no changes in the  correspond control property too. Or we talk here only about `INotifyPropertyChanged` derived objects?

But from **API Usage** section we clearly see that you test _control to object_ scenario, which doesn't make sense in any thread other then UI.

**Comment by zeh-almeida** [comment]
I am in the process of building a better example as I only have my code to show for, which wasn't crafted for this case specifically.

Yes, we are talking about `INotifyPropertyChanged` derived objects, specifically when you have a asynchronous operation on said object which triggers the event.

[I have a previous example here](https://github.com/zeh-almeida/6502-sharp/commit/6216195368990d09160d29056632c2c6cbec3a13) from my project, which illustrates how I was creating the binding.

However, because of [my message](https://github.com/zeh-almeida/6502-sharp/blob/ce309b7998ccad660be160cf51599ca1fcd9ce43/Cpu.MVVM/CpuModel.cs#L52) is asynchronous, some of the message handlers changed their properties and the binding raised the `System.InvalidOperationException`.

Hope this clarifies the issue better.

**Comment by kirsan31** [comment]
Thank you, now are all clear.

About your proposal.
Personally, I think this is not good approach (pass param to constructor).  I will try to argue my position.
- For now we know that all methods of `Binding` class must be used in UI thread only.
- After you change we can obtain two version of  `Binding` class object - UI only (old behavior) and cross thread with `Invoke`. This will lead to confusion in the code - it will not be clear when you can use `Binding` not from the main thread, but when you cannot.
- Also it's no way (other then reflection) to determinate in runtime is our `Binding` instance cross thread or not. May be to add some property?
- Why `Invoke` and not `BeginInvoke` or able to select? Did you analyze possibilities of deadlocks scenarios in case of `Invoke`?

Overall I think approach suggested by @RussKie is more suitable here...

**Comment by zeh-almeida** [comment]
I agree, my proposal is probably not the best approach, I only tested using the scenario I found, without analyzing possibilites of deadlocks or other adverse situations.

I used `Control.Invoke` instead of `Control.BeginInvoke` in this case because the existing code calls the `property.SetValue` directly and I wanted to avoid making changes to existing behavior, [as you can see here](https://github.com/dotnet/winforms/blob/64f34e0f829342e0777ffe216c6fe920682bc800/src/System.Windows.Forms/src/System/Windows/Forms/Binding.cs#L1010).

I have no problem in changing the proposal to add a property toggle for this behavior instead of changing the constructor, as I now understand that changing any signature could be a problem for the designer and other tools.

I just want to make sure I understand the suggestion correctly:
I should add a `public bool UseControlInvoke {get; set;}` property to the `Binding` class and, if true, it should execute the `property.SetValue` method in the `Control.Invoke` action, removing the need to change the constructor.

Is this correct?

**Comment by kirsan31** [comment]
> I used Control.Invoke instead of Control.BeginInvoke in this case because the existing code calls the property.SetValue directly and I wanted to avoid making changes to existing behavior, [as you can see here](https://github.com/dotnet/winforms/blob/64f34e0f829342e0777ffe216c6fe920682bc800/src/System.Windows.Forms/src/System/Windows/Forms/Binding.cs#L1010).

Yea `BeginInvoke` will require more investment. But in case of cross thread operation, I think in 90% cases, logically, you will want to use `BeginInvoke` instead of `Invoke`.

> I just want to make sure I understand the suggestion correctly:
> I should add a `public bool UseControlInvoke {get; set;}` property to the `Binding` class and, if true, it should execute the `property.SetValue` method in the `Control.Invoke` action, removing the need to change the constructor.

It seems to me that here we are stepping on thin ice of cross thread UI control🙄 Imho ideal solution will be something like ``BindingCT`` or even `BindingAsync` 😯
Lets wait for response from maintainers...

**Comment by RussKie** [comment]
Since the current behaviour is sync the fix would be to use Invoke. BeginInvoke or adding an async variant would be a new behaviour.
________________________________

**Comment by KlausLoeffelmann** [comment]
Just a quick update:
We're still very much considering this but need to address a couple of higher priority issues, first.

**Comment by elachlan** [comment]
@KlausLoeffelmann did you want to re-prioritize this?

**Comment by elachlan** [comment]
Related: #8532

**Comment by KlausLoeffelmann** [comment]
Well, I want to get it done for 9. 
I am not sure exactly when I get to it.
